Fullscreen video does not always match window size at certain resolutions

mwyrzykowski opened the pull request:
#### c819e77ac5671e6c2df6822462a0294871b492c5
<pre>
Fullscreen video does not always match window size at certain resolutions
<a href="https://bugs.webkit.org/show_bug.cgi?id=258484">https://bugs.webkit.org/show_bug.cgi?id=258484</a>
rdar://110724034

Reviewed by Jer Noble.

The video is using ObjectFit::Contain and UIKit is trying to preserve the initial aspect ratio
from when the video entered fullscreen during resizes. However, we requested the the window to
only resize to integral point values, so the aspect ratio can be off more after a resize than it was initially.

* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
(-[WebAVPlayerLayer calculateTargetVideoFrame]):
Ensure the target video frame is integral values.

* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::areAspectRatiosEssentiallyEqual):
At most the difference is one point, i.e., deviceScaleFactor, so we can use this as a tolerance
value passed to WTF::areEssentiallyEqual.

(WebCore::RenderVideo::videoBox const):
(WebCore::RenderVideo::updatePlayer):
Make the video take up the entire area if the aspect ratio is almost identical and we are in fullscreen.

Canonical link: <a href="https://commits.webkit.org/265672@main">https://commits.webkit.org/265672@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94caea7397700eb7a76ebafe610988e742b26d6f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11345 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11895 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10812 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11366 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11539 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13722 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12381 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9602 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13413 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9676 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10281 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17475 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10745 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10438 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13654 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8980 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10029 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2787 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10712 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/rendering/RenderVideo.cpp:
```diff
@@ -175,6 +175,11 @@ void RenderVideo::imageChanged(WrappedImagePtr newImage, const IntRect* rect)
 
+static bool areAspectRatiosEssentiallyEqual(const auto& size1, const auto& size2)
+{
+    return fabs(size1.aspectRatio() - size2.aspectRatio()) <= 0.01f;
```
I know you explained in the commit message why this is reasonable for comparing aspect ratios, but would WTF::areEssentiallyEqual work here instead?

mwyrzykowski replied:
Oh thanks I didn't know about `WTF::areEssentiallyEqual`. I could pass an epsilon to that as the default of `std::numeric_limits<T>::epsilon()` is too small.

mwyrzykowski replied:
I updated the PR to use WTF::areEssentiallyEqual and compute the tolerance based on the deviceScaleFactor and the content size of the layer in fullscreen. This is a more accurate check.

mwyrzykowski committed:
Fullscreen video does not always match window size at certain resolutions
https://bugs.webkit.org/show_bug.cgi?id=258484
rdar://110724034

Reviewed by Jer Noble.

The video is using ObjectFit::Contain and UIKit is trying to preserve the initial aspect ratio
from when the video entered fullscreen during resizes. However, we requested the the window to
only resize to integral point values, so the aspect ratio can be off more after a resize than it was initially.

* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
(-[WebAVPlayerLayer calculateTargetVideoFrame]):
Ensure the target video frame is integral values.

* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::areAspectRatiosEssentiallyEqual):
At most the difference is one point, i.e., deviceScaleFactor, so we can use this as a tolerance
value passed to WTF::areEssentiallyEqual.

(WebCore::RenderVideo::videoBox const):
(WebCore::RenderVideo::updatePlayer):
Make the video take up the entire area if the aspect ratio is almost identical and we are in fullscreen.

Canonical link: https://commits.webkit.org/265672@main
